[docs] Adds EAS Update preview docs (#14455)
expo/expo: docs/pages/preview/eas-update/deployment-patterns.md

---
title: Patterns to deploy your app
---

Once we've created features and fixed bugs, we want to get our changes out to our users as soon as we can, as safely as we can. Often "safe" and "fast" are opposing forces when delivering code to our users. We could push our code directly to production, which would be fast, but it would also be unsafe since we never tested our code. On the other hand, we could make test builds, share them with a QA team, and release periodically. That would be safer, but also slower to deliver changes to our users.

Depending on your project, you'll have some tolerance for how "fast" and how "safe" you'll need to be when delivering updates to your users.

> There are a three parts that make up a successful deployment process that we'll need to consider:

Reviewer: ```suggestion
There are three parts that make up a successful deployment process that we'll need to consider:
```

1. Creating builds

>    1. We can create builds for production use only.
>    2. We can create builds for production use and separate builds for testing use.

Reviewer: ```suggestion
   a. We can create builds for production use only.
   b. We can create builds for production use and separate builds for testing use.
```

2. Testing changes
   1. We can test changes with TestFlight and Play Store Internal Track.
   2. We can test changes with an internal distribution build.
   3. We can test changes with Expo Go or a development app.
3. Publishing updates
   1. We can publish updates without considering branch names.

>    2. We can create update branches that are environment based, like `production` and `staging`.

Reviewer: Everywhere else the names of branches and channels are quoted instead of code-formatted

>    2. We can create update branches that are environment based, like `production` and `staging`.
>    3. We can create update branches that are version based, like `version-1.0`, which enables us to promote updates from one channel to another.

Reviewer: ```suggestion
   2. We can create update branches that are environment-based, like `production` and `staging`.
   3. We can create update branches that are version-based, like `version-1.0`, which enables us to promote updates from one channel to another.
```

## Two-command flow

This flow is the simplest and fastest flow, with the fewest amount of safety checks. It's great for trying out Expo and for smaller projects. Here are the parts of the deployment process above that make up this flow:

Creating builds: (a) Create builds for production use only.

Testing changes: (c) Test changes with Expo Go or a development app.

Publishing updates: (a) Publish updates without considering branch names.

Diagram of flow:

![Two command deployment diagram](/static/images/eas-update/deployment-two-command.png)

Explanation of flow:

- Develop a project locally and test changes in Expo Go.
- Run `eas build` to create builds, then submit them to app stores. These builds are for public use, and should be submitted/reviewed, and released on the app stores.
- When we have updates we'd like to deliver, run `eas branch:publish production` to deliver updates to our users immediately.

Advantages of this flow:

- This flow does not require bookkeeping extra version or environment names, which makes it easy to communicate to others.
- Delivering updates to builds is very fast.

Disadvantages of this flow:

> - There are no pre-production checks to make sure the code will function as we want. We can test with Expo Go or a development app, but this is less safe than having a dedicated testing environment.

Author: This flow assumes that there's only one branch, "production". To test out an update before it goes live, we can either use Expo Go, or a development app (custom dev client). We could also make an internal distribution build with the new update code included in the build and test that build, but that would add more steps to this flow. The goal of this flow is to be the simplest, fastest, and most dangerous flow for developers. It's really great for developers who are getting started and who don't need tons of QA before going live.

## Branch flow

This flow focuses on de-risking your deployments and preserving previous versions of your app. It's great for medium- to large-sized projects. Here are the parts of the deployment process above that make up this flow:

Creating builds: (b) Create builds for production use and separate builds for testing use.

Testing changes: (a) Test changes on TestFlight and the Play Store Internal Track and/or (b) Test changes with internal distribution builds.

Publishing updates: (c) create update branches that are version based, like `version-1.0`.

Diagram of flow:

![Branch deployment diagram](/static/images/eas-update/deployment-branch.png)

Explanation of flow:

- Develop a project locally and test changes in Expo Go.
- Create builds with channels named `production` , which will eventually get reviewed and become available on app stores. Create another set of builds with channels named `staging`, which will be used for testing on TestFlight and the Play Store Internal Track.

> - Set up expo-github-action to publish updates when merging commits to branches.

Reviewer: ```suggestion
- Set up `expo-github-action` to publish updates when merging commits to branches.
```

- Merge changes into a branch named `version-1.0`.

> - Use the website UI or CLI to point the `staging` channel at the EAS Update branch `version-1.0`. Test the update by opening the apps on TestFlight and the Play Store Internal Track.
> - When ready, use the website UI or CLI to point the `production` channel at the EAS Update branch `version-1.0`.
> - Then, create another GitHub branch named `version-1.1`, and merge commits until the new features and fixes are ready. When they are, use the website UI or CLI to point the a `staging` channel at the EAS Update branch `version-1.1`.
> - When ready, use the website UI or CLI to point the `production` channel at the EAS Update branch `version-1.1`.

Reviewer: ```suggestion
- Use the website or EAS CLI to point the `staging` channel at the EAS Update branch `version-1.0`. Test the update by opening the apps on TestFlight and the Play Store Internal Track.
- When ready, use the website or EAS CLI to point the `production` channel at the EAS Update branch `version-1.0`.
- Then, create another GitHub branch named `version-1.1`, and merge commits until the new features and fixes are ready. When they are, use the website or EAS CLI to point the `staging` channel at the EAS Update branch `version-1.1`.
- When ready, use the website or EAS CLI to point the `production` channel at the EAS Update branch `version-1.1`.
```

Disadvantages of this flow:

- Bookkeeping of branch names is required for this flow, which will mean communicating with your team which branches are currently pointed at your test builds and your production builds.

## Persistent staging flow

This flow is a slight variation of the "Branch flow". The main difference is that it does not require bookkeeping branches. Instead, we'll have a persistent `staging` and `production` branches we can merge into forever. It's great for medium- projects to large-sized projects. Here are the parts of the deployment process above that make up this flow:

Creating builds: (b) Create builds for production use and separate builds for testing use.

Testing changes: (a) Test changes on TestFlight and the Play Store Internal Track and/or (b) Test changes with internal distribution builds.

Publishing updates: (c) create update branches that are environment based, like `staging` and `production`.

Diagram of flow:

![Staging deployment diagram](/static/images/eas-update/deployment-staging.png)

Explanation of flow:

- Develop a project locally and test changes in Expo Go.

> - Create builds with channels named `production` , which will eventually get reviewed and become available on app stores. Create another set of builds with channels named `staging`, which will be used for testing on TestFlight and the Play Store Internal Track.

Reviewer: ```suggestion
- Create builds with channels named `production`, which will eventually get reviewed and become available on app stores. Create another set of builds with channels named `staging`, which will be used for testing on TestFlight and the Play Store Internal Track.
```

> - Set up expo-github-action to publish updates when merging commits to branches.

Reviewer: ```suggestion
- Set up `expo-github-action` to publish updates when merging commits to branches.
```

- Merge changes into a branch named `staging`. The GitHub Action will publish an update and make it available on our test builds.
- When ready, merge changes into the `production` branch to publish an update to our production builds.

Advantages of this flow:

- This flow is an even balance between safety and speed.
- It's easy to communicate to your team, since deploying updates occurs when merging into GitHub branches named `staging` and `production`.

Disadvantages of this flow:

- Checking out previous versions of your app is slightly more complex, since we'd need to check out an old commit instead of an old branch.
- When merging to `production`, the update would be re-built and re-published instead of moved from the builds with channel `staging` to the builds with channel `production`.

## Platform specific flow

> This flow is for projects that need to build and update their iOS and Android apps separately all the time. It will result in separate commands for delivering updates to the iOS and Android apps. Here are the parts of the deployment process above that make up this flow:

Reviewer: ```suggestion
This flow is for projects that need to build and update their Android and iOS apps separately all the time. It will result in separate commands for delivering updates to the Android and iOS apps. Here are the parts of the deployment process above that make up this flow:
```

Creating builds: (a) Create builds for production use only, or (b) create builds for production use and separate builds for testing use.

Testing changes: (a) Test changes on TestFlight and the Play Store Internal Track and/or (b) Test changes with internal distribution builds.

Publishing updates: (c) create update branches that are environment- and platform-based, like `ios-staging`, `ios-production`, `android-staging`, and `android-production`.

Diagram of flow:

![Platform specific deployment diagram](/static/images/eas-update/deployment-platform-specific.png)

Explanation of flow:

- Develop a project locally and test changes in Expo Go.

> - Create builds with channels named like `ios-staging`, `ios-production`, `android-staging`, and `android-production`. Then put the `ios-staging` build on TestFlight and submit the `ios-production` build to the public App Store. Likewise, put the `android-staging` build on the Play Store Internal Track, and submit the `android-production` build to the public Play Store.

Reviewer: Everywhere else (?) channel names are double-quoted instead of code-formatted:
```suggestion
- Create builds with channels named like "ios-staging", "ios-production", "android-staging", and "android-production". Then put the "ios-staging" build on TestFlight and submit the "ios-production" build to the public App Store. Likewise, put the "android-staging" build on the Play Store Internal Track, and submit the "android-production" build to the public Play Store.
```

> - Set up expo-github-action to publish updates when merging commits to branches.

Reviewer: ```suggestion
- Set up `expo-github-action` to publish updates when merging commits to branches.
```
